src/type/validate.js

```diff
@@ -578,6 +581,64 @@ function validateInputFields(
   });
 }
 
+function validateInputObjectCircularReferences(
+  context: SchemaValidationContext,
+): void {
+  // Modified copy of algorithm from 'src/validation/rules/NoFragmentCycles.js'.
+  // Tracks already visited types to maintain O(N) and to ensure that cycles
+  // are not redundantly reported.
+  const visitedTypes = Object.create(null);
+
+  // Array of types nodes used to produce meaningful errors
+  const fieldPath = [];
+
+  // Position in the type path
+  const fieldPathIndexByTypeName = Object.create(null);
+
+  const typeMap = context.schema.getTypeMap();
+  for (const type of objectValues(typeMap)) {
```

Comment:
It's bad that we need to iterate over every type one more time.
Would be great to integrate this check into the loop inside `validateTypes` function without affecting its readability.